instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/triggers/ApiGatewayRestTrigger.java

```diff
@@ -0,0 +1,197 @@
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanKindExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanStatusBuilder;
+import io.opentelemetry.instrumentation.awslambdacore.v1_0.AwsLambdaRequest;
+import io.opentelemetry.instrumentation.awslambdacore.v1_0.internal.Trigger;
+import io.opentelemetry.instrumentation.awslambdacore.v1_0.internal.TriggerMismatchException;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes.FaasTriggerValues;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public final class ApiGatewayRestTrigger extends Trigger {
+
+  @Override
+  public boolean matches(AwsLambdaRequest request) {
+    return request.getInput() instanceof APIGatewayProxyRequestEvent;
+  }
+
+  @Override
+  public String extract(AwsLambdaRequest request) {
+    APIGatewayProxyRequestEvent req = requireCorrectEventType(request);
+    return req.getResource();
+  }
+
+  @Override
+  public void extract(SpanStatusBuilder spanStatusBuilder, AwsLambdaRequest request,
+      @Nullable Object response, @Nullable Throwable error) {
+    APIGatewayProxyResponseEvent res = requireCorrectResponseType(response);
+    int statusCodeInt = res.getStatusCode();
+    if (statusCodeInt >= 400 && statusCodeInt < 600) {
+      spanStatusBuilder.setStatus(StatusCode.ERROR);
+    } else {
+      spanStatusBuilder.setStatus(StatusCode.UNSET);
+    }
+  }
+
+  @Override
+  public void onStart(AttributesBuilder attributes, Context parentContext,
+      AwsLambdaRequest request) {
+
+    // TODO consider using some things from io.opentelemetry.instrumentation.api.instrumenter.http
+    try {
+      APIGatewayProxyRequestEvent req = requireCorrectEventType(request);
+      ProxyRequestContext requestContext = req.getRequestContext();
+      RequestIdentity identity = requestContext.getIdentity();
+      Map<String, String> headers = req.getHeaders();
+
+      attributes.put("test.null.attribute", (String) null);
+      attributes.put(FAAS_TRIGGER, FaasTriggerValues.HTTP);
+      attributes.put(FAAS_TRIGGER_TYPE, "Api Gateway Rest");
+      attributes.put(HTTP_METHOD, requestContext.getHttpMethod());
+      attributes.put(HTTP_ROUTE, requestContext.getResourcePath());
+      attributes.put(HTTP_URL, getHttpUrl(req, lowercaseMap(request.getHeaders())));
+      attributes.put(NET_SOCK_PEER_ADDR,
+          identity.getSourceIp()); // TODO change this in python and nodejs
+      attributes.put(USER_AGENT_ORIGINAL,
+          headers.get("User-Agent")); // TODO change this in python and nodejs
+      attributes.put(HTTP_SCHEME, headers.get("X-Forwarded-For"));
```

Comment:
Would not you get ip from `X-Forwarded-For` not http scheme?

Reply:
🤦 It was supposed to be `X-Forwarded-Proto`